src/accountsdb/buffer_pool.zig

```diff
@@ -0,0 +1,1174 @@
+/// handles - frames may outlive the associated rc. Frames are considered dead
+/// when are evicted (which may only happen with 0 open handles).
+/// A frame is always alive when it can be found in frame_map; a frame is always
+/// dead when it is in free_list.
+/// A frame dies when its index is evicted from HierarchicalFifo (inside of
+/// evictUnusedFrame).
+pub const BufferPool = struct {
+    pub const FrameMap = std.AutoHashMapUnmanaged(FileIdFileOffset, FrameIndex);
+
+    /// indices of all free frames
+    /// free frames have a refcount of 0 *and* have been evicted
+    free_list: AtomicStack(FrameIndex),
+
+    /// uniquely identifies a frame
+    /// for finding your wanted index
+    // TODO: thread safety?
+    frame_map: FrameMap,
+
+    frames: []Frame,
+    frames_metadata: FramesMetadata,
+
+    /// used for eviction to free less popular (rc=0) frames first
+    eviction_lfu: HierarchicalFIFO,
+
+    /// NOTE: we might want this to be a threadlocal for best performance? I don't think this field is threadsafe
+    io_uring: if (builtin.os.tag == .linux) std.os.linux.IoUring else void,
+
+    pub fn init(
+        init_allocator: std.mem.Allocator,
+        num_frames: u32,
+    ) !BufferPool {
+        if (num_frames == 0 or num_frames == 1) return error.InvalidArgument;
+
+        const frames = try init_allocator.alignedAlloc(Frame, std.mem.page_size, num_frames);
+        errdefer init_allocator.free(frames);
+
+        var frames_metadata = try FramesMetadata.init(init_allocator, num_frames);
+        errdefer frames_metadata.deinit(init_allocator);
+
+        var free_list = try AtomicStack(FrameIndex).init(init_allocator, num_frames);
+        errdefer free_list.deinit(init_allocator);
+        for (0..num_frames) |i| free_list.appendAssumeCapacity(@intCast(i));
+
+        var io_uring = if (builtin.os.tag == .linux) blk: {
+            // NOTE: this is pretty much a guess, maybe worth tweaking?
+            // think this is a bit on the high end, libxev uses 256
+            const io_uring_entries = 4096;
+
+            break :blk try std.os.linux.IoUring.init(
+                io_uring_entries,
+                0,
+            );
+        } else {};
+        errdefer if (builtin.os.tag == .linux) io_uring.deinit();
+
+        var frame_map: FrameMap = .{};
+        try frame_map.ensureTotalCapacity(init_allocator, num_frames);
+        errdefer frame_map.deinit(init_allocator);
+
+        return .{
+            .frames = frames,
+            .frames_metadata = frames_metadata,
+            .free_list = free_list,
+            .frame_map = frame_map,
+            .eviction_lfu = try HierarchicalFIFO.init(init_allocator, num_frames / 10, num_frames),
+            .io_uring = io_uring,
+        };
+    }
+
+    pub fn deinit(self: *BufferPool, init_allocator: std.mem.Allocator) void {
+        init_allocator.free(self.frames);
+        self.frames_metadata.deinit(init_allocator);
+        if (builtin.os.tag == .linux) {
```

Comment:
should have the same logic with `use_io_uring`

Reply:
fixed in b3aa41d4c7a8e90beeeed1d917bbf2b2c374183c